[WIP] New variadic arguments implementation

ChrisJefferson opened the pull request:
Changes the notation for variadic argument lists from `(a,b,arg)` (where arg is a special name) to `(a,b, ...p)` (where `p` can be any identifier, including `arg`). This doesn't change functionality at all, it just means that we don't break any code which was using `arg` as an argument name.

This patch leaves the older (new) notation in for now, if we like this I'll remove it and fix the standard library.

For the curious, after all this stuff is finished, all GAP 4.7 code will behave identically, this will only add new notation.

markuspf commented:
Just for the record, two slightly alternative syntax variants were discussed, namely using

```
f := function(a,b,args...) ... end;
```
or
```
f := function(a,b,args[]) ... end;
```

All variants are similarly easy to implement. My (not very strong) order of preference is `args[]` > `args...` > `...args`.

stevelinton commented:
I would favour the x[] syntax as being more obviously related to concepts already in GAP, rather than concepts from Python or C++.

alex-konovalov commented:
My order of preferences would be:
* `f(a, b, x[])`
* `f(a, b, x...)`
* `f(a, b, ...x)`

I like `f(a, b, x[])` the most because it conveys the fact that `x` will be a list whenever I will see `x` in the function body much better than the other two. 

My second choice wins over my third choice because `f(a, b, ...x)` separates variadic argument `x` from other arguments more visibly than `f(a, b, x...)` what's also good. And speculating further, accidental cutting of `x` in `...x` will result in syntactically invalid code, while accidental cutting of dots in `x...` will produce syntactically correct but wrong code.

The last one has also a nice point though: the brain may expand `f(a, b, x...)` into `f(a, b, x1, x2, x3, ...)` easier than `f(a, b, ...x)`. But since my preference is the 1st one, I will stop here.

rbehrends commented:
For what it's worth (i.e. my two Eurocents, give or take :) ), I favor the `x...` option, because it seems to most closely approximate conventional mathematical notation; the `...x` notation feels like it's the wrong way around. I'm not particularly enthusiastic about the `x[]` notation (which makes it look like an expression, not a declaration).

fingolfin commented:
My two cents:  I favor  `x...` over `...x` for the reasons @rbehrends gave; and I dislike `x[]` -- but any of them is vastly better than breaking compatibility with existing code.

markuspf commented:
As I mentioned before, I will not stand in the way of any choice that is made.

fingolfin commented:
@alex-konovalov wrote:
> I like f(a, b, x[]) the most because it conveys the fact that x will be a list whenever I will see x in the function body much better than the other two. 

That is actually one of the reasons why I dislike it: It gives the incorrect impression that this is about marking an argument as receiving a list; but this is just a byproduct, after all, other arguments may also receive lists. The true intent here is to mark the argument as one receiving extra variadic arguments...

In particular, I am concerned that it might mislead people with experiences in other languages which have similar `x[]` syntax. So somebody with a background in C/C++ might think they *have* to use [] to mark arguments that expect a list. They would also expect it to "work" for arbitrary arguments, and would not guess that it actually is used for variadic arguments. The `...` OTOH is familar to those people. So for these people, the `x[]` syntax might be a bad choice.

Now, there are of course lots of other people who are not familiar with either notation, and thus would not find either more or less confusing. I believe most users would in fact be rather neutral about the choice of syntax. So, I have a group of users for which `x[]` might end up negative, and a group where it is neutral. Hence in total, it seems inferior to me compared to `x...` or `...x`..  :-)

I also still don't get why you are concerned about people "cutting" their code accidentally. If that happens, you are always screwed. Moreover, I actually don't understand your specific example: if I botch a copy & paste of `f(a, b, x...)` the way you describe it, I'd end up with ``f(a, b, x` and that actually is also syntactically invalid. Of course a novice programmer might end up "fixing" that by adding a `)`. But are you then also concerned about a novice programmer copy & pasting
```
if x then
   funcA();
   funcB();
fi;
```
as
```
if x then
   funcA();
```
and then adding `fi;`, accidentally changing the meaning of the code?

alex-konovalov commented:
Well, perhaps if someone uses coding style
```
something:=function(
                                   a, b, c...
                                 )
```
and cuts dots, then one could still get a valid input. But I've never seen such. I am not fixated on this at all - I think is something to take into account only if you have two choices which seem absolutely identical otherwise. You can not predict all ways to shoot yourself in a foot in advance. As I said, I will be happy with any of three choices. Seems to me like we are leaning towards `...x` at the moment, and Steve has a good point about `..` instead of `...`

fingolfin commented:
Here is another (weak) reason why I personally prefer `...` over `[]`: It fits nicely with another syntax I'd like to see, and mentioned on the mailing list, inspired by Julia: Allow writing
``` func(1,2,3,list...)```
or
``` func(1,2,3,...list)```
as synonym for
```CallFuncList(func, Concatenation([1,2,3], list));```
which I would find highly useful. Of course we could still do that, even if we do not use `...` or `..` for variadic arguments. And we could also use a syntax like
`````` func(1,2,3,list[])```
but that would then be ambiguous for the parser, and feels contrived.

markuspf commented:
I start leaning towards `args...` for multiple reasons. I still won't be in any implementation's way though.

alex-konovalov commented:
And I start leaning towards `*x` but also will be happy with anything.

BTW, one should check readability of the code both WITH spaces and WITHOUT spaces. To help you compare all six options suggested so far:

```
f(a,b,c,*x)
f(a,b,c,x[])
f(a,b,c,x..)
f(a,b,c,..x)
f(a,b,c,x...)
f(a,b,c,...x)
```
and
```
f( a, b, c, *x )
f( a, b, c, x[] )
f( a, b, c, x.. )
f( a, b, c, ..x )
f( a, b, c, x... )
f( a, b, c, ...x )
```

alex-konovalov commented:
Please vote: details at https://mail.gap-system.org/pipermail/gap/2015-December/000356.html

fingolfin commented:
Looking at the poll so far, it seems that "x..." is the clear "winner" so far. @ChrisJefferson would you mind rewriting this PR to use that syntax?

fingolfin commented on src/read.c:
```diff
@@ -1181,7 +1184,13 @@ void ReadFuncExpr (
+        if (isvarg) {
+            SyntaxError("Only final argument can be variadic");
+        }
+    
```
The indention here is different from the surrounding code (which seems to mix tabs and spaces). I think it is best to match the indenting mode of the surrounding code, to keep the code readable. For now, that is (we can fix the indention globally later on, perhaps during GAP days).

fingolfin commented on src/read.c:
```diff
@@ -1248,15 +1267,16 @@ void ReadFuncExpr (
-    if (narg >= 1 && ! strcmp( "arg", CSTR_STRING( ELM_LIST(nams, narg) ) ) )
+    /* 'function( a,b, ...p )' takes a variable number of arguments           */
+    /* Also, we special case function(arg)                                   */
+    if (isvarg || ( narg >= 1 && ! strcmp( "arg", CSTR_STRING( ELM_LIST(nams, narg) ) )) )
```
I guess the `narg >= 1` should be changed to `narg == 1` here, so that only the old `function(arg)` leads to a variadic function, while `function(a,b,c,arg)` is (again, as in the past) a simple 4 parameter function.

fingolfin commented on src/read.c:
```diff
@@ -1248,15 +1267,16 @@ void ReadFuncExpr (
+    if ( narg > 1 && ! strcmp( "arg", CSTR_STRING( ELM_LIST(nams, narg) ) ) ) {
+      		SyntaxWarning("arg more arguments");
+      }
     /*     if ( narg == 1 && ! strcmp( "arg", CSTR_STRING( ELM_LIST(nams,1) ) ) )
```
I guess this is just in here to flag places that already switched to the new-but-now-hopefully-soon-deprecated variadic `function(a,b,c,arg)`? If so, perhaps the message should be adjusted? Or perhaps we can remove it?
In particular, do we know of any users of this feature outside of the library (i.e. any packages)?

ChrisJefferson commented:
Hopefully now cleaned up.

In the end, x... was the clear winner, so gone with that.

Removed `function(x, arg)` style variadic arguments. `function(arg)` still works. Added tests, and made sure the parser gives sensible messages when users confuse `..` and `...`.

No documentation yet.

stevelinton commented:
Did you change the (few) examples of function (x, arg) that I had already put in?

If not I can.

Steve

> On 15 Dec 2015, at 12:49, Christopher Jefferson <notifications@github.com> wrote:
> 
> Hopefully now cleaned up.
> 
> In the end, x... was the clear winner, so gone with that.
> 
> Removed function(x, arg) style variadic arguments. function(arg) still works. Added tests, and made sure the parser gives sensible messages when users confuse .. and ....
> 
> No documentation yet.
> 
> —
> Reply to this email directly or view it on GitHub.
>

ChrisJefferson commented:
Yup (although I just noticed there was a couple in the docs, so I'm doing that now).

stevelinton commented:
One consequence of this that we MUST do a second 4.8 beta

stevelinton commented:
Thanks
> On 15 Dec 2015, at 12:59, Christopher Jefferson <notifications@github.com> wrote:
> 
> Yup (although I just noticed there was a couple in the docs, so I'm doing that now).
> 
> —
> Reply to this email directly or view it on GitHub.
>

markuspf commented:
I think if @fingolfin is happy with this, we should go ahead and merge, and get the next beta out so we can do 4.8.

fingolfin commented on doc/ref/language.xml:
```diff
@@ -1735,15 +1735,14 @@ only needs about <C>Log(<A>n</A>)</C> steps.
-and this list is assigned to the new variable corresponding to the formal
-argument <C>arg</C>. There are  two  typical  scenarios  for  wanting  such  a
+and this list is assigned to the new variable corresponding to the final formal
+argument There are  two  typical  scenarios  for  wanting  such  a
```
Missing a full stop here.

fingolfin commented on doc/ref/language.xml:
```diff
@@ -1735,15 +1735,14 @@ only needs about <C>Log(<A>n</A>)</C> steps.
 As noted in Section&nbsp;<Ref Sect="Function Calls"/>,
-the case where a functions last formal argument has the name
-<C>arg</C>, is special.
+the case where a function's last argument is followed by <C>...</C> is special.
```
I guess that means section "Function Calls" also needs to be adapted (yep, around line 785)

fingolfin commented on doc/ref/language.xml:
```diff
@@ -1804,12 +1800,16 @@ The user should compare the above with the  &GAP; function <Ref Func="Sum"/>
-The argument <C>arg</C> when used as the single argument name of some function
+The <C>...</C> notation when used as the single argument name of some function
 <A>f</A> tells &GAP; that when it encounters <A>f</A> that it should form  a  list
 out of the arguments of <A>f</A>.
```
That sentence is weird in the revised form... It doesn't really make sense to only talk about the "single argument" case here, right?

fingolfin commented on lib/function.g:
```diff
@@ -191,8 +191,9 @@ DeclareOperationKernel( "SetNameFunction", [IS_OBJECT, IS_STRING], SET_NAME_FUNC
-##  the number returned is - the total number of parameters including the <C>arg</C>.
+##  For functions that use <C>...</C> or <C>arg</C> to take a variable number
+##  of arguments,
+##  the number returned is - the total number of parameters.
```
Why is this on three lines, instead of three?

BTW, minor nitpick: I find the text "... is - the total number" quite confusing (I realize that it was there before, but I already found it confusing then). How about changing it to "... is -1 times the total number", while we are at it?

fingolfin commented on src/read.c:
```diff
@@ -1175,11 +1181,19 @@ void ReadFuncExpr (
-	      }
+	    if (isvarg) {
+	        SyntaxError("Only final argument can be variadic");
+	    }
```
Very minor point: Right now, one gets this:
```
gap> function(x..,y) end;
Syntax error: Three dots required for variadic argument list
function(x..,y) end;
           ^
```
That might be slightly confusing / misleading, so can we change this to print "only the final argument can be variadic" instead? If not resp. if this is a lot of work, then just ignore this -- I am merely wondering≥

ChrisJefferson replied:
That would be a bit of fiddling unfortunately, as we don't catch this isn't the last argument until the next loop around, but we need to make sure we reject the two dots as ever being a variadic function. I agree the message isn't great, I'll look at it.

fingolfin replied:
Well, don't waste too much time on it; changing this is certainly not a priority, just "would be nice to have". But there are many other things that would be nice to have, and many of them would be much, much nicer (such as what we get by merging this PR ... ;-).

fingolfin commented on src/read.c:
```diff
@@ -1175,11 +1181,19 @@ void ReadFuncExpr (
+
+	if(TLS(Symbol) == S_DOTDOT) {
+	    SyntaxError("Three dots required for variadic argument list");
+	}
```
Off-topic, but: Why do we sometimes start SyntaxError messages in upper case, and sometimes in lowercase?

fingolfin commented on src/read.c:
```diff
@@ -1248,15 +1274,13 @@ void ReadFuncExpr (
 
-    /* 'function( ... arg )' takes a variable number of arguments              */
-    if (narg >= 1 && ! strcmp( "arg", CSTR_STRING( ELM_LIST(nams, narg) ) ) )
+    /* 'function( a,b, ...p )' takes a variable number of arguments           */
```
That comment still says `...p`, should be `p...`

fingolfin commented on src/read.c:
```diff
@@ -1248,15 +1274,13 @@ void ReadFuncExpr (
 	narg = -narg;
       }
-	
+      
```
If you change whitespace here, perhaps just remove them (and leave an empty line)?

fingolfin commented on tst/testinstall/varargs.tst:
```diff
@@ -0,0 +1,54 @@
+gap> START_TEST("varargs.tst");
+gap> f := function(a,b...) return[a,b]; end;
```
Super pedantic nitpick (sorry): Insert a space before `[a,b]`?

fingolfin commented on tst/testinstall/varargs.tst:
```diff
@@ -0,0 +1,54 @@
+gap> Print(f,"\n");
+function ( a... )
+    return a;
+end
```
Again pedantic, but: why are the tests for `function(arg) return arg; end;` and `f := function(a...) return a; end;` doing different things, in a different order? Ideally, the two functions should be equivalent (except for the output of `Print`), so I'd expect them to have equivalent / equal tests.

fingolfin commented on tst/testinstall/varargs.tst:
```diff
@@ -0,0 +1,54 @@
+[ 1, [ 2, 3 ] ]
+gap> f := function(arg) return arg; end;
+function( arg... ) ... end
+gap> Print(f,"\n");
```
You could also just write `Display(f);` :-)

fingolfin commented on tst/testinstall/varargs.tst:
```diff
@@ -0,0 +1,54 @@
+gap> f := function(a,b..) end;
+Syntax error: Three dots required for variadic argument list in stream line 1
+f := function(a,b..) end;
+                  ^
```
Maybe also add (some of) these?
```
gap> function(a,b..) end;
Syntax error: Three dots required for variadic argument list
function(a,b..) end;
             ^
gap> function(a...,b) end;
Syntax error: Only final argument can be variadic
function(a...,b) end;
             ^
gap> function(a..,b) end;
Syntax error: Three dots required for variadic argument list
function(a..,b) end;
           ^
gap> function(a....,b) end;
Syntax error: ) expected
function(a....,b) end;
             ^
gap> function(a,b....) end;
Syntax error: ) expected
function(a,b....) end;
               ^
```

fingolfin commented on lib/function.g:
```diff
@@ -566,8 +565,12 @@ InstallMethod( ViewObj, "for a function", true, [IsFunction], 0,
+        isvarg := true;
         narg := -narg;
     fi;
+    if narg = 1 and nams[1] = "arg" then
```
Turns out this check can fail: Apparently is possible for nams to be `fail`, when looking at C level code. The resulting function, when printed, is shown like this:
```
function ( <<arg-1>> )
    <<kernel or compiled code>>
end
```
A reproducible (but not yet minimal) test case:
```
METHODS_OPERATION(Socle,1);
```

ChrisJefferson commented:
Now updated, with @fingolfin 's many comments fixed.

The problem of printing functions without argument lists is fixed, but not in the .tst files (as I didn't track down a small function for testing yet).

fingolfin commented:
Good job, Chris, thanks!

alex-konovalov commented:
The diffs below occur in the test of manual examples - I will fix this:
```
########> Diff in [ "./../../lib/function.g", 361, 376 ]
# Input is:
PrintNumberFromDigits := function ( arg )
   CallFuncList( Print, arg );
   Print( "\n" );
  end;
# Expected output:
function( arg ) ... end
# But found:
function( arg... ) ... end
########
########> Diff in [ "./../../lib/function.g", 361, 376 ]
# Input is:
PrintDigits := function ( arg )
   Print( arg );
   Print( "\n" );
  end;
# Expected output:
function( arg ) ... end
# But found:
function( arg... ) ... end
########
########> Diff in [ "./../../lib/function.g", 399, 406 ]
# Input is:
f:=ReturnTrue;  
# Expected output:
function( arg ) ... end
# But found:
function( arg... ) ... end
########
########> Diff in [ "./../../lib/function.g", 426, 433 ]
# Input is:
f:=ReturnFalse;  
# Expected output:
function( arg ) ... end
# But found:
function( arg... ) ... end
########
########> Diff in [ "./../../lib/function.g", 453, 460 ]
# Input is:
oops:=ReturnFail;  
# Expected output:
function( arg ) ... end
# But found:
function( arg... ) ... end
########
########> Diff in [ "./../../lib/function.g", 479, 484 ]
# Input is:
n:=ReturnNothing;
# Expected output:
function( object ) ... end
# But found:
function( object... ) ... end
########
########> Diff in [ "./../../lib/function.g", 504, 513 ]
# Input is:
f:=ReturnFirst;
# Expected output:
function( object ) ... end
# But found:
function( object... ) ... end
########
########> Diff in [ "./mloop.xml", 624, 642 ]
# Input is:
ErrorNoTraceBack := function(arg) # arg is special variable that GAP
                                    # knows to treat as list of arg's
    local SavedOnBreak, ENTBOnBreak;
    SavedOnBreak := OnBreak;        # save current value of OnBreak

    ENTBOnBreak := function()       # our `local' OnBreak
    local s;
      for s in arg do
        Print(s);
      od;
      OnBreak := SavedOnBreak;      # restore OnBreak afterwards
    end;

    OnBreak := ENTBOnBreak;
    Error();
  end;
# Expected output:
function( arg ) ... end
# But found:
function( arg... ) ... end
########
########> Diff in [ "./../../lib/teaching.g", 288, 294 ]
# Input is:
isbntest:=CheckDigitTestFunction(10,11,[1,2,3,4,5,6,7,8,9,-1]); 
# Expected output:
function( arg ) ... end
# But found:
function( arg... ) ... end
########
```
